### PR TITLE
#1296 Update unit testing way

### LINKED ISSUE #566: Do not use Tests in CMAKE_BUILD_TYPE
CMAKE_BUILD_TYPE should not be set to Tests, it is not its purpose in CMake

### LINKED ISSUE #1296: Standardize unit testing
Remove the use of non standard unit testing cmake build types Tests and TestsCoverage. Use cmake options instead

### PR DESCRIPTION
**Pull request by lecourtoisflo**
Standardizing the cmake cmake testing reveals that the old ways
had side effects. some warnings (seen as errors) were not raised

fixes #1296 
fixes #566

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -42,7 +42,6 @@ jobs:
       image: dynawo/dynawo-ci:latest
       options: --user root
     env:
-      DYNAWO_BUILD_TYPE: TestCoverage
```
Comment (gautierbureau) [inline .github/workflows/ci.yml:45]: DYNAWO_BUILD_TYPE is mandatory, should be Debug. Then DYNAWO_BUILD_TESTS_COVERAGE should be define to ON

**Review by gautierbureau — CHANGES_REQUESTED** [review]

**Comment by gautierbureau** [comment]
Now that DYNAWO_BUILD_TYPE can only be Debug or Release we should get rid of DYNAWO_BUILD_TYPE_THIRD_PARTY that was defined only to handle the case of Tests and TestsCoverage values

File: dynawo/CMakeLists.txt [hunk 2]
```diff
@@ -21,6 +21,9 @@ if(${CMAKE_VERSION} VERSION_GREATER "3.15.0")
   endif()
 endif()
 
+option(DYNAWO_BUILD_TESTS "Choose to build the unit tests")
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/CMakeLists.txt:24]: OK

[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/CMakeLists.txt:24]: With the script, it is a environment variable as well

**Comment by lecourtoisflo** [comment]
> Now that DYNAWO_BUILD_TYPE can only be Debug or Release we should get rid of DYNAWO_BUILD_TYPE_THIRD_PARTY that was defined only to handle the case of Tests and TestsCoverage values

Agree

File: dynawo/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/CMakeLists.txt:24]: Also, in most template cmake I found, the project name is used as a prefix for all options

File: util/envDynawo.sh [hunk 3]
```diff
@@ -763,18 +775,14 @@ config_dynawo() {
   if [ $DYNAWO_LIBARCHIVE_HOME_DEFAULT != true ]; then
     CMAKE_OPTIONAL="$CMAKE_OPTIONAL -DLIBARCHIVE_HOME=$DYNAWO_LIBARCHIVE_HOME"
   fi
-  case $DYNAWO_BUILD_TYPE in
-    Tests|TestCoverage)
+  if [ $DYNAWO_BUILD_TESTS = ON -o $DYNAWO_BUILD_TESTS_COVERAGE = ON ]; then
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply util/envDynawo.sh:778]: OK

File: dynawo/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/CMakeLists.txt:24]: In cmake DYNAWO_BUILD_TESTS is a cache variable and in envDynawo.sh it is an env variable. I prefer the prefix in the env variables to be able to easily recognize which variables were set by dynawo in your envronment, in cmake the prefix is not useful as it is only local to cmake and cannot interfer with anything else. All options are not prefixed in cmake as it is not useful at all

[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/CMakeLists.txt:24]: OK

**Review by gautierbureau — APPROVED** [review]

File: dynawo/sources/Common/test/Test.cpp [hunk 4]
```diff
@@ -430,7 +430,7 @@ TEST(CommonTest, testSparseMatrix) {
     smj4.addTerm(i, i);
   }
   for (int icol = 0; icol < smj4.nbCol(); ++icol) {
-    for (int ind = smj4.Ap_[icol]; ind < smj4.Ap_[icol + 1]; ++ind) {
+    for (int ind = smj4.Ap_[icol]; ind < static_cast<int>(smj4.Ap_[icol + 1]); ++ind) {
```
Comment (rosiereflo) [inline dynawo/sources/Common/test/Test.cpp:433]: Wouldn't it be better to replace the int index with an unsigned?

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
 /**
  * @file Modeler/DataInterface/test/TestIIDMModels.cpp
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp:15]: Please update the wrong file name comment (Unit tests for Criteria)

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp [hunk 6]
```diff
@@ -13,7 +13,7 @@
 
 /**
  * @file Modeler/DataInterface/test/TestIIDMModels.cpp
- * @brief Unit tests for DataInterface/*IIDM classes
+ * @brief Unit tests for DataInterfaceIIDM classes
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp:16]: Same as above

File: util/envDynawo.sh [hunk 7]
```diff
@@ -644,7 +646,7 @@ config_3rd_party() {
   if [ $DYNAWO_LIBARCHIVE_HOME_DEFAULT != true ]; then
     CMAKE_OPTIONAL="$CMAKE_OPTIONAL -DLIBARCHIVE_HOME=$DYNAWO_LIBARCHIVE_HOME"
   fi
-  case $DYNAWO_BUILD_TYPE_THIRD_PARTY in
+  case $DYNAWO_BUILD_TYPE in
```
Comment (rosiereflo) [inline util/envDynawo.sh:649]: Why not testing DYNAWO_BUILD_TESTS?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: util/envDynawo.sh (hunk 7 above)
Comment (gautierbureau) [reply util/envDynawo.sh:649]: yes test should be similar as the one in config_dynawo

Comment (lecourtoisflo) [reply util/envDynawo.sh:649]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp:15]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp (hunk 6 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp:16]: OK

File: dynawo/sources/Common/test/Test.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply dynawo/sources/Common/test/Test.cpp:433]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp [hunk 8]
```diff
@@ -12,7 +12,7 @@
 //
 
 /**
- * @file Modeler/DataInterface/test/TestIIDMModels.cpp
+ * @file Modeler/DataInterface/IIDM/test/TestCriteria.cpp
  * @brief Unit tests for DataInterfaceIIDM classes
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp:16]: Unit tests for Criteria classes

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp [hunk 9]
```diff
@@ -12,7 +12,7 @@
 //
 
 /**
- * @file Modeler/DataInterface/test/TestIIDMModels.cpp
+ * @file Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp
  * @brief Unit tests for DataInterfaceIIDM classes
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp:16]: Unit tests for Static Parameters classes

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp (hunk 8 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestCriteria.cpp:16]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp (hunk 9 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestIIDMModelsStaticParameters.cpp:16]: OK

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**